pylint_django/augmentations/__init__.py

```diff
@@ -826,10 +807,6 @@ def apply_augmentations(linter):
     # Method could be a function (get, post)
     suppress_message(linter, ClassChecker.leave_functiondef, 'no-self-use',
                      is_model_view_subclass_method_shouldnt_be_function)
-    # Unused argument 'request' (get, post)
-    suppress_message(linter, VariablesChecker.leave_functiondef, 'unused-argument',
-                     is_model_view_subclass_unused_argument)
-    suppress_message(linter, VariablesChecker.leave_functiondef, 'unused-argument', is_argument_named_request)
```

Comment:
I have tested an implementation based on monkey patching `_check_unused_arguments`.  Unfortunately this would result into incorrect `redefined-outer-name` warnings for `request`. Pylint generates `redefined-outer-name` warnings for function arguments unless these argument names are ignored through `ignored-argument-names` ([see `VariablesChecker.visit_functiondef`](https://github.com/PyCQA/pylint/blob/96b3892506de6033ecb4f6e288bd4f5b211d894c/pylint/checkers/variables.py#L827)). I have added a test case for this.

The implementation in the latest commit now wraps `VariablesChecker._is_name_ignored` method.

This gives us the following advantages:
- Integrated in the existing augmentation structure
- No messing around with appending something to the `ignored-argument-names` regex and the possible edge cases thereof. Don't know if appending `|request` is sufficient for all cases.

Overall I think that this is the cleaner implementation _except of monkey patching the private method_. What do you think?

If we want to go forward with this I would like to refactor wrapping of `VariablesChecker.leave_module` and `VariablesChecker._is_name_ignored` into a separate function `apply_wrapped_augmentations` (or something). I would like to do this to gather the context information ("must update in thread safe way") in one location.

Reply:
+1 for the refactor.

I'm not entirely sure I understand the comments about `redefined-outer-name` but overall PR looks good

Reply:
Refactoring finished. 

Concerning the thread safe monkey patching:
I am not so sure that the `if current_is_name_ignored.__name__ == '_is_name_ignored':` is actually needed anymore. After https://github.com/PyCQA/pylint-django/commit/3e4fc69f27d644f4ab4c86f7f995903bef368361 the `__name__` of the wrapped object shouldn't change anymore (because of ` @functools.wraps`). Therefore, currently, the wrapping always happens.
I have tested this by printing `VariablesChecker._is_name_ignored .__name__` after the if and running pylint with `-j4`. The output was always `_is_name_ignored`. Same for `leave_module`.

Additionally pylint creates new subprocesses and therefore this whole "updating in a thread safe manner" shouldn't be necessary either way, I think? For every new process the object has to be monkey patched again.

We can leave it like it is and there shouldn't be any problems, because it is already like this as of 5 months ago. Its just a bit confusing and apparently unnecessary.

Reply:
Regarding `redefined-outer-name`:

Here is an example for pylints behaviour regarding `ignored-argument-names` and `redefined-outer-name` warnings to (hopefully) make it clear:

File `pylintrc.example`:
``` ini
[VARIABLES]
ignored-argument-names=user_id
```

File `example.py`
``` python
# pylint: disable=missing-docstring
from django.http import JsonResponse

user_id = None  # pylint: disable=invalid-name


def user_detail(request, user_id):  # [unused-argument]
    # nothing is done with user_id
    return JsonResponse({'username': 'steve'})
```

Running pylint with default settings:
```
$ pylint --load-plugins=pylint_django example.py
************* Module example
example.py:7:25: W0621: Redefining name 'user_id' from outer scope (line 4) (redefined-outer-name)
example.py:7:25: W0613: Unused argument 'user_id' (unused-argument)

------------------------------------------------------------------
Your code has been rated at 5.00/10 (previous run: 5.00/10, +0.00)
```

Running pylint with the config file `pylintrc.example` from above:
```
$ pylint --rcfile pylintrc.example --load-plugins=pylint_django example.py                                                               4

-------------------------------------------------------------------
Your code has been rated at 10.00/10 (previous run: 5.00/10, +5.00)
```

No `redefined-outer-name` warning is generated for `user_id`.

Reply:
Thanks for your example. I'm still not convinced I follow your thoughts but TBH I've had a lot on my plate recently and I'm totally incapable of understanding this ATM. I guess I need to go back and read how this area of pylint works and understand things on my own before I can begin understanding your examples.

I see the test suite passes so we should be good. I will merge this PR as-is and push out a new version so other folks don't have to wait on this bug.

If you feel cleaning up the code more and exploring/giving more examples around redefined-outer-name feel free to do so in a separate pull request. Otherwise we'll fix it when it breaks!